tianshou/env/vecenv.py

```diff
@@ -180,6 +180,101 @@ def close(self) -> List[Any]:
+    def __init__(self, env_fns: List[Callable[[], gym.Env]],
+                 wait_num: Optional[int] = None) -> None:
+        """
+        :param wait_num: used in asynchronous simulation if the time cost of
+            ``env.step`` varies with time and synchronously waiting for all
+            environments to finish a step is time-wasting. In that case, we
+            can return when ``wait_num`` environments finish a step and keep
+            on simulation in these environments. If ``None``, asynchronous
+            simulation is disabled; else, ``1 <= wait_num <= env_num``.
+        """
+        super().__init__(env_fns)
+        self.wait_num = wait_num or len(env_fns)
+        assert 1 <= self.wait_num <= len(env_fns), \
+            f'wait_num should be in [1, {len(env_fns)}], but got {wait_num}'
+        self.waiting_conn = []
+        # environments in self.ready_id is actually ready
+        # but environments in self.waiting_id are just waiting when checked,
+        # and they may be ready now, but this is not known until we check it
+        # in the step() function
+        self.waiting_id = []
+        # all environments are ready in the beginning
+        self.ready_id = list(range(self.env_num))
+
+    def reset(self, id: Optional[Union[int, List[int]]] = None) -> np.ndarray:
+        if id is None:
+            id = range(self.env_num)
+            # reset all environments
+            self.ready_id = list(range(self.env_num))
+        elif np.isscalar(id):
+            assert id in self.ready_id, f'can only reset ' \
+                                        f'ready environments {self.ready_id}'
+            id = [id]
+        for i in id:
+            assert i not in self.waiting_id, f'cannot reset environment {i}' \
+                                             f' which is stepping now!'
+        return super().reset(id)
+
+    def render(self, **kwargs) -> List[Any]:
+        if len(self.waiting_id) > 0:
+            raise RuntimeError(
+                f"environments {self.waiting_id} are still "
+                f"stepping, cannot render now")
+        return super().render(**kwargs)
+
+    def close(self) -> List[Any]:
+        if self.closed:
+            return []
+        # finish remaining steps, and close
+        self.step(None)
+        return super().close()
+
+    def step(self,
+             action: Optional[np.ndarray],
+             id: Optional[Union[int, List[int]]] = None
+             ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Provide the given action to the environments. For the first call,
+        action should be the action for each environment; for the consecutive
+        call, the action sequence should correspond to the ``env_id`` filed
+        in the last returned ``info``
+        If action is None, fetch unfinished step() calls instead.
+        """
+        if id is not None:
+            raise ValueError("cannot specify the id of environments"
```

Comment:
I would rather accept `id` argument to be a subset of : "``env_id`` filled in the last returned ``info``"

Reply:
There are two alternatives, one is the current code, the other is as you said, let id be a subset of env_id filled in the last returned info. In the latter case, ``Collector.collect`` would need to pass this argument again and again, and I don't think this has practical significance. The whole point of asynchronous simulation is to step those ready environments as soon as possible, so I take the former alternative, assume the observation returned in the last step all yield corresponding action.

If you have some usecase of the latter alternative, please tell me.

Reply:
I see your point. Indeed it is not intended to be used in `Collector.collect`. But I like to use `Tianshou.data` for many stuffs far from RL. Part of my work is simulation VS reality analysis. For this, I'm not using any collector. Just raw Batch and VectorEnv. Then I run simulations by hand step by step to carefully analyse what is happening. 

To wrap up, try to think outside the box. Tianshou is desgined for RL, but part of it goes far beyond. Especially Batch and VecEnv. Making them as versatile as possible may lead to unexpected yet very interesting use cases, from which Tianshou could get better if we seize the opportunity.

Reply:
That's reasonable, suggestion taken!